### DISCUSSION TITLE
There should be a way to make fixtures return **copies** of their return values

### DISCUSSION THREAD

**Discussion by alexander-soare** [pytest-dev/pytest#10761] [Ideas] [answered]
(This could also be a question, as I'm a relatively new Pytest user)

The other day I discovered that a module scoped fixture's return value is mutable by tests:

```python
@pytest.fixture(scope='module')
def fixture():
    print("Running fixture")
    return [0]

def test_A(fixture):
    print("A:", fixture)  # prints[0]
    fixture[0] += 1

def test_B(fixture):
    print("B: ", fixture)  # prints [1]
```

I'm not sure this is always going to be desirable behavior. Shouldn't there be an option to make sure that the return value of the fixture is copied on request? Not doing so requires developers to be vigilant, which doesn't feel like a scalable solution.

**Comment by RonnyPfannschmidt** [comment, marked as answer]
It's the intended behaviour for larger scopes

If you need fresh per test use function scope

**Reply by alexander-soare** [reply]
What if my fixture takes a long time to run? I might want to just copy the return value each time.

**Reply by RonnyPfannschmidt** [reply]
So have a session scope fixtures to compute and a function scope one to make the copy

**Reply by alexander-soare** [reply]
Great! Thank you.
